### PR TITLE
Create E0774

### PR DESCRIPTION
**Pull request by GuillaumeGomez**

### CODE REVIEW

File: compiler/rustc_error_codes/src/error_codes/E0774.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+```
+#[derive(Clone)] // ok!
+struct Bar {
+    field: u32,
+}
+```
+
+You can find more information about `derive` in the [Rust Book].
+
+[Rust Book]: https://doc.rust-lang.org/1.30.0/book/second-edition/appendix-03-derivable-traits.html
```
Comment (pickfire) [inline compiler/rustc_error_codes/src/error_codes/E0774.md:24]: Do we need to link to `1.30.0` which is quite old?

Comment (GuillaumeGomez) [reply compiler/rustc_error_codes/src/error_codes/E0774.md:24]: Try this page with current stable (it's super weird though).

**Comment by pickfire** [comment]
> https://doc.rust-lang.org/1.30.0/book/second-edition/appendix-03-derivable-traits.html

Weird, the local version of rust book does show it. Why doesn't the [online version](https://doc.rust-lang.org/1.46.0/book/second-edition/appendix-03-derivable-traits.html) does?

file:///home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/share/doc/rust/html/book/appendix-03-derivable-traits.html

File: compiler/rustc_error_codes/src/error_codes/E0774.md (hunk 1 above)
Comment (pickfire) [reply compiler/rustc_error_codes/src/error_codes/E0774.md:24]: > https://doc.rust-lang.org/1.30.0/book/second-edition/appendix-03-derivable-traits.html

Weird, the local version of rust book does show it. Why doesn't the online version does?

file:///home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/share/doc/rust/html/book/appendix-03-derivable-traits.html

Comment (jyn514) [reply compiler/rustc_error_codes/src/error_codes/E0774.md:24]: Works fine on stable: https://doc.rust-lang.org/book/appendix-03-derivable-traits.html

File: compiler/rustc_expand/src/expand.rs [hunk 2]
```diff
@@ -529,9 +529,11 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
...
         let attr = self.cx.sess.find_by_name(item.attrs(), sym::derive);
         let span = attr.map_or(item.span(), |attr| attr.span);
-        let mut err = self
-            .cx
-            .struct_span_err(span, "`derive` may only be applied to structs, enums and unions");
+        let mut err = self.cx.struct_span_err_with_code(
+            span,
+            "`derive` may only be applied to structs, enums and unions",
+            DiagnosticId::Error("E0774".into()),
+        );
```
Comment (jyn514) [inline compiler/rustc_expand/src/expand.rs:536]: Maybe use `struct_span_err!` instead? That also makes sure `E0774` is a valid error code IRRC.

File: compiler/rustc_expand/src/base.rs [hunk 3]
```diff
@@ -1032,6 +1032,15 @@ impl<'a> ExtCtxt<'a> {
...
     }
 
+    pub fn struct_span_err_with_code<S: Into<MultiSpan>>(
+        &self,
+        sp: S,
+        msg: &str,
+        code: DiagnosticId,
+    ) -> DiagnosticBuilder<'a> {
+        self.sess.parse_sess.span_diagnostic.struct_span_err_with_code(sp, msg, code)
+    }
```
Comment (jyn514) [inline compiler/rustc_expand/src/base.rs:1042]: Is this necessary?

Comment (pickfire) [reply compiler/rustc_expand/src/base.rs:1042]: I think it is good for feature parity with `rustc_errors`.

Comment (pickfire) [reply compiler/rustc_expand/src/base.rs:1042]: The only thing I don't ilke it it adds yet another `struct_span_err_with_code` to the docs, which make the original version harder to find.

Comment (jyn514) [reply compiler/rustc_expand/src/base.rs:1042]: I don't see why it needs to be added to `rustc_expand` directly though. You could just use `struct_span_err!` which makes this nice and easy to type: `struct_span_err!(self.sess, sp, msg, E0774)`.

Comment (pickfire) [reply compiler/rustc_expand/src/base.rs:1042]: Maybe we should remove `struct_span_err` above later?

**Comment by GuillaumeGomez** [comment]
@jyn514 @pickfire Did you decide on a solution? Either way I'm fine with it. :)

**Comment by jyn514** [comment]
I think this should use `struct_span_err!`.

**Comment by GuillaumeGomez** [comment]
Then let's go!

**Comment by GuillaumeGomez** [comment]
Updated!

File: compiler/rustc_expand/src/expand.rs [hunk 4]
```diff
@@ -529,9 +529,12 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
     fn error_derive_forbidden_on_non_adt(&self, derives: &[Path], item: &Annotatable) {
         let attr = self.cx.sess.find_by_name(item.attrs(), sym::derive);
         let span = attr.map_or(item.span(), |attr| attr.span);
-        let mut err = self
-            .cx
-            .struct_span_err(span, "`derive` may only be applied to structs, enums and unions");
+        let mut err = rustc_errors::struct_span_err!(
```
Comment (jyn514) [inline compiler/rustc_expand/src/expand.rs:532]: nit: this is generally imported and used without the `rustc_errors` prefix

**Review by jyn514 — APPROVED** [review]

File: compiler/rustc_expand/src/base.rs (hunk 3 above)
Comment (pickfire) [reply compiler/rustc_expand/src/base.rs]: I am fine with either but if we follow `struct_span_err!` can we remove `struct_span_err` that was already there in favor of `struct_span_err!` later on?

Comment (GuillaumeGomez) [reply compiler/rustc_expand/src/base.rs]: I'm opening an issue for that. I think it's better to do it in another PR.

**Pull request merged**